### PR TITLE
Handle targets provided in deps as a result of GoSource.resolver

### PR DESCRIPTION
**Pull request by dzbarsky**
**What type of PR is this?**
Starlark fixup

**What does this PR do? Why is it needed?**
Make the migration process a little smoother

**Which issues(s) does this PR fix?**
https://github.com/bazelbuild/rules_go/pull/4030#issuecomment-2310491357

**Other notes for review**

### CODE REVIEW

**Comment by dzbarsky** [comment]
@tyler-french @fmeum wdyt?

File: go/private/context.bzl [hunk 1]
```diff
@@ -311,6 +311,17 @@ def _library_to_source(go, attr, library, coverage_instrumented):
     if library.resolve:
         library.resolve(go, attr, source, _merge_embed)
 
+        # TODO(zbarsky): Remove this once downstream has a chance to migrate.
+        if verify_resolver_deps:
+            has_targets = False
+            for dep in source["deps"]:
+                if type(dep) != "struct":
```
Comment (fmeum) [inline go/private/context.bzl:318]:
```suggestion
                if type(dep) == "Target":
```
seems more direct

File: go/private/context.bzl [hunk 2]
```diff
@@ -311,6 +311,17 @@ def _library_to_source(go, attr, library, coverage_instrumented):
...
+                    has_targets = True
+                    break
+            if has_targets:
+                print("""Detected Targets in `source["deps"]` as a result of _resolver. Please provide `GoArchive`s instead. This will be an error in the future.""")
```
Comment (fmeum) [inline go/private/context.bzl:322]: Maybe add a snippet describing how to do that (`[GoArchive]`)?

**Review by fmeum — COMMENTED** [review]
I like it!

File: go/private/context.bzl (hunk 2 above)
Comment (dzbarsky) [reply go/private/context.bzl]: yep, good idea

**Review by tyler-french — APPROVED** [review]

**Comment by tyler-french** [comment]
This is working, thank you!

**Comment by dzbarsky** [comment]
> This is working, thank you!

Awesome! any other issues in Uber repo?

**Comment by tyler-french** [comment]
Sure, I'll try to grab that tomorrow.

And I think everything is mostly good other than some nogo issues I haven't gotten around to.

**Review by fmeum — APPROVED** [review]

**Pull request merged**